Preliminary Giga Touch support on Zephyr

KurtE opened the pull request:
Zephyr: Added in our preliminary support for touch on the display

In ArduinoCore-zephyr added a callback to register with the input class.  When set it simply forwards the callbacks to this function.

Then added some support.  We still don't have things like gestures as I don't think the zephyr input class has support for it.

Have a simple touch paint sketch (now supporting multiple touches, that appears to work

Depends on:
- [x] https://github.com/arduino-libraries/Arduino_GigaDisplayTouch/pull/15
- [x] https://github.com/arduino/ArduinoCore-zephyr/pull/165

KurtE commented:
@per1234 @leonardocavagnis and all...

Not sure what to do with this clang format failure?  Looks like you added this something like two years ago, and 
no one ever actually ran it on the existing code and examples and updated github. 

Not sure that is something that should be done as part of this PR?  As for example, I have not touched any of your
examples. 

Thanks
Kurt

iabdalkader reviewed (changes requested):
I haven't tested this yet but I think we should move this to a separate file? For example `src/Arduino_GigaDisplayTouchZephyr.cpp` and then move the old one to `src/Arduino_GigaDisplayTouchMbed.cpp`

Also, please fix your formatting. For example:

- Please pick a style:
```
Arduino_GigaDisplayTouch::~Arduino_GigaDisplayTouch()
{
}

bool Arduino_GigaDisplayTouch::begin() {
```

- Fix indentation:
```C
void Arduino_GigaDisplayTouch::end()
{
        registerGigaTouchCallback(nullptr);
}
```

- Remove empty lines such as:
```C
uint8_t Arduino_GigaDisplayTouch::getTouchPoints(GDTpoint_t* points, uint32_t timeout) {

    // First wait to see if we get any events.    
```

- Remove empty spaces at end of lines:
```C
    // First wait to see if we get any events.    
```

iabdalkader commented on src/Arduino_GigaDisplayTouch.h:
```diff
@@ -77,19 +172,19 @@ class Arduino_GigaDisplayTouch {
-      Arduino_GigaDisplayTouch(TwoWire& wire, 
-                               uint8_t intPin, 
-                               uint8_t rstPin, 
+    #else
```
Do we need to duplicate the class?
```suggestion
    #elif defined(__ZEPHYR__)
    Arduino_GigaDisplayTouch();
    #else
```

KurtE replied:
Note: I pushed up a clang-format of the files and then restricted these changes to the zephyr area.

This is sort of a global question, that we (@mjs513 and I) have been trying to ask. How does Arduino want to handle some of these libraries?  There are several different options:

a) Completely different libraries, that potentially part of the Board installs.  Like: Camera, Wire, SPI are done now

b) As libraries you install from the library manager.  With this there are at least 3 variations that I have seen (and done)
1) Try to intersperse all the code the MBED and Zephyr stuff with lots of #ifdef...
2) Like I did, where there is like 1 #ifdef or the like with the two different implementations.
3) like 2, except that split the files up. like you might have Arduio_GigaDisplayTouch_zephyr.cpp and these have
the same #if stuff like 2, but easier to locate the code and defines.

There are pros and cons to all of these.

So again not sure which way?

iabdalkader replied:
I don't think there's a single rule that fits all, this is probably best decided on a case-by-case basis, so please use your better judgement. For example, if you find yourself adding a whole new alternate implementation, then it should probably go in a new file. However, there might be an argument for keeping everything in the same file if support for Mbed and other cores is planned to be removed, that would make it easier to do so. I'm also not sure if this is planned or not, perhaps @pennam knows more about this.

EDIT: No plan to deprecate other cores anytime soon, so the header file should use interspersed `ifdef`'s, the implementation should be in two separate files: `Arduino_GigaDisplayTouchZephyr.cpp` and `Arduino_GigaDisplayTouchMbed.cpp`.

> Completely different libraries, that potentially part of the Board installs. Like: Camera, Wire, SPI are done now

I think core libraries provide core functionality like Wire and SPI. Libraries like touchscreen, BLE etc.. are not core functionality.

iabdalkader commented on src/Arduino_GigaDisplayTouch.h:
```diff
@@ -122,7 +217,7 @@ class Arduino_GigaDisplayTouch {
       TwoWire&          _wire;
       uint8_t           _intPin;
-      mbed::InterruptIn _irqInt; 
+      mbed::InterruptIn _irqInt;
```
Same here we just hide all these private functions and variables with `#ifdef __MBED__`.

KurtE commented:
@iabdalkader - I did a combination, where I tried to merge the two sections of the header file into one.
I then split out the zephyr implementation into it's own .cpp file.

Everything was run through the clang-format 
I built both examples on MBED.  Fixed that they both gave warnings on order of initialization. 

Both run.  However the IRQ version dies in MBED after a few touches.  
```
++ MbedOS Error Info ++
Error Status: 0x80FF0144 Code: 324 Module: 255
Error Message: Assertion failed: id
Location: 0x8048FE1
File: Event.h+207
Error Value: 0x0
Current Thread: main Id: 0x24003200 Entry: 0x80451AD StackSize: 0x8000 StackMem: 0x24004698 SP: 0x2407FEEC
For more info, visit: https://mbed.com/s/error?error=0x80FF0144&osver=61700&core=0x411FC271&comp=2&ver=90200&tgt=GIGA
-- MbedOS Error Info --

```
Not related to any changes I made as did not change anything, except the order as I mentioned and it 
failed before that change as well.

Let me know what you think.

KurtE commented:
Here is a test sketch that can do the touchpaint that I am showing in the ArduinoCore-zephyr PR.  I modified it such that
it now works on both MBED and zephyr.

[zephyr_GIGA_shield_touchpaint-250812a.zip](https://github.com/user-attachments/files/21743328/zephyr_GIGA_shield_touchpaint-250812a.zip)

iabdalkader commented on src/Arduino_GigaDisplayTouch.h:
```diff
@@ -52,87 +54,123 @@ typedef struct GDTpoint_s GDTpoint_t;
+#endif
+#elif defined(__ZEPHYR__)
+  Arduino_GigaDisplayTouch();
+#endif
```
```suggestion
/**
 * @brief Construct a new touch controller for Giga Display Shield.
 *
 * @param wire A reference to the Wire interface to be used for communication
 * with the touch controller.
 * @param intPin The interrupt pin number for the touch controller.
 * @param rstPin The reset pin number for the touch controller.
 * @param addr The device address for the touch controller.
 */
#if defined(__ZEPHYR__)
  Arduino_GigaDisplayTouch();
#elif defined(ARDUINO_GIGA)
  Arduino_GigaDisplayTouch(TwoWire &wire = Wire1,
                           uint8_t intPin = PinNameToIndex(PI_1),
                           uint8_t rstPin = PinNameToIndex(PI_2),
                           uint8_t addr = GT911_I2C_ADDR_BA_BB);
#elif defined(ARDUINO_PORTENTA_H7_M7)
  Arduino_GigaDisplayTouch(TwoWire &wire = Wire,
                           uint8_t intPin = PinNameToIndex(PD_4),
                           uint8_t rstPin = PinNameToIndex(PD_5),
                           uint8_t addr = GT911_I2C_ADDR_BA_BB);
#else
  Arduino_GigaDisplayTouch(TwoWire &wire, uint8_t intPin, uint8_t rstPin,
                           uint8_t addr);
#endif
```

KurtE replied:
I pulled in your changes and then rebuilt for both mbed and zephyr and fixed any issues I found

iabdalkader commented on src/Arduino_GigaDisplayTouch.h:
```diff
@@ -52,87 +54,123 @@ typedef struct GDTpoint_s GDTpoint_t;
+   * @param timeout in microseconds.
+   * @return uint8_t The number of detected touch points.
+   */
+  uint8_t getTouchPoints(GDTpoint_t *points, uint32_t timeout = 0);
```
The API should be the same for both. 
```suggestion
  /**
   * @brief Check if a touch event is detected and get the touch points.
   * @param points The array containing the coordinates of the touch points.
   * @return uint8_t The number of detected touch points.
   */
  uint8_t getTouchPoints(GDTpoint_t *points);
```

iabdalkader commented on src/Arduino_GigaDisplayTouch.h:
```diff
@@ -52,87 +54,123 @@ typedef struct GDTpoint_s GDTpoint_t;
+  uint8_t _gt911ReadOp(uint16_t reg, uint8_t *data, uint8_t len);
+  void _gt911onIrq();
+  uint8_t _gt911ReadInputCoord(uint8_t *pointsbuf, uint8_t &contacts);
+#elif defined(__ZEPHYR__)
+  /*
+   * @brief Support for callbacks from zephyr
+   */
+  typedef struct {
+    size_t x;
+    size_t y;
+    bool pressed;
+  } touch_point_t;
+
+  static touch_point_t
+      _zephyr_touch_points[CONFIG_INPUT_GT911_MAX_TOUCH_POINTS];
+
+  static uint8_t _zephyr_touch_cb_slot_num;
+  static struct k_sem _zephyr_touch_event_sync;
+
+  static void _touch_event_callback(struct input_event *evt, void *user_data);
+
```
All of this can be moved to the CPP file.
```suggestion
```

KurtE replied:
done

iabdalkader reviewed (changes requested):
* I had many comments on the cpp file, so I thought it would just be easier to send you an updated file. See the attached file.
* Again, please use the name suggested in my first review:
`Arduino_GigaDisplayTouchZephyr.cpp`

* I've already formatted the library and renamed the mbed implementation in https://github.com/arduino-libraries/Arduino_GigaDisplayTouch/pull/15, so please revert any unrelated changes to minimize your diff.

[Arduino_GigaDisplayTouchZephyr.cpp.txt](https://github.com/user-attachments/files/21751879/Arduino_GigaDisplayTouchZephyr.cpp.txt)

KurtE committed:
Zephyr: Added in our preliminary support for touch on the display

Zephyr: Added in our preliminary support for touch on the display

In ArduinoCore-zephyr added a callback to register with the
input class.  When set it simply forwards the callbacks to
this function.

Then added some support.  We still don't have things like
gestures as I don't think the zephyr input class has support
for it.

Have a simple touch paint sketch (now supporting multiple
touches, that appears to work

Edit: The header file now combines the two different
versions of the class into one.

Added #ifdef __MBED__ in the mbed implementation file so that
the code would not compile when we are doing zephyr builds

KurtE commented:
I believe I have all of your changes pulled in now.  I rebased off of your updated main branch, and then
add in your changes.  Had to do a few fixes, like #ifdef __MBED__ in the mbed version so that it
would not build on zephyr.

I have also updated to your changes on the ArduinoCore-zephyr rebuilt and ran with this updated version

iabdalkader commented:
Please fix the formatting issues, see https://github.com/arduino-libraries/Arduino_GigaDisplayTouch/actions/runs/16943340161/job/48017964802?pr=14

Otherwise, looks good.

iabdalkader commented:
@KurtE  Please fix the minor formatting issue reported in https://github.com/arduino-libraries/Arduino_GigaDisplayTouch/actions/runs/16943340161/job/48017964802?pr=14

We'll merge this after the core PR.

KurtE commented:
> @KurtE Please fix the minor formatting issue reported in https://github.com/arduino-libraries/Arduino_GigaDisplayTouch/actions/runs/16943340161/job/48017964802?pr=14
> 
> We'll merge this after the core PR.

I reran the clang-format on it... The only thing it changed was to indent the the last line *** end of file ****
hopefully the online version will be happy

iabdalkader commented on src/Arduino_GigaDisplayTouchZephyr.cpp:
```diff
@@ -0,0 +1,103 @@
+
+void Arduino_GigaDisplayTouch::onDetect(void (*handler)(uint8_t,
+                                                        GDTpoint_t *)) {
+  UNUSED(handler);
```
@KurtE Should we implement this while we're at it? @pennam suggests we do so.

iabdalkader commented:
I'll work on adding the missing IRQ mode. Thanks @KurtE for adding this.
